### PR TITLE
Fix tok per sec

### PR DESCRIPTION
**Pull request by nathan-az**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

#### Changelog
What are the changes made in this PR?
* Adds an attribute to ParallelDims to count non-dp workers
* Adjusts tokens per second calculation to adjust for over-counting of non-dp workers

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/full_finetune_distributed.py [hunk 1]
```diff
@@ -965,7 +965,9 @@ def train(self) -> None:
                                     else self._optim_ckpt_wrapper
                                 ),
                             ),
-                            "tokens_per_second_per_gpu": num_tokens
+                            "tokens_per_second_per_gpu": (
+                                num_tokens / self.parallel_dims.non_data_parallel_size
+                            )
                             / (time_per_step * self.world_size),
```
Comment (joecummings) [inline recipes/full_finetune_distributed.py:971]: Couldn't we just divide by self.dp_size here instead ?

Comment (nathan-az) [reply recipes/full_finetune_distributed.py:971]: Hmm I'm not following. An example to explain my motivation (and elucidate any misunderstanding on my behalf).

Suppose we have batch size 1, with 1000 tokens in the sample, and TP dim=4.

Due to the all-reduce, this will report as 4000 tokens. But functionally, each device processed only 250 tokens.

* `num_tokens / world_size=1000` over-reports due to the TP workers
* `num_tokens / dp_size` would report 4000 tokens (since dp size is 1) - over-reporting *twice* by the TP workers.

The motivation behind my adjustment is:
* In the denominator, dividing by world size gets us to the actual number of GPUs
* In the numerator, the adjustment corrects the number of tokens actually processed during the update step

Comment (joecummings) [reply recipes/full_finetune_distributed.py:971]: Okay, I see - thank you for your very clear thinking!

**Review by joecummings — APPROVED** [review]
👍
